addons/point_of_sale/static/src/app/models/product_product.js

```diff
@@ -0,0 +1,224 @@
+    async _onScaleNotAvailable() {}
+
+    isConfigurable() {
+        const attrByIds = this.models["product.attribute"].getAllBy("id");
+        return this.attribute_line_ids.some((line) => line.attribute_id?.id in attrByIds);
+    }
+
+    get isScaleAvailable() {
+        return true;
+    }
+
+    get parentCategoryIds() {
+        const categories = [];
+        let category = this.categ_id;
+
+        while (category) {
+            categories.push(category.id);
+            category = category.parent_id;
+        }
+
+        return categories;
+    }
+
+    get parentPosCategIds() {
+        const current = [];
+        const categories = this.pos_categ_ids;
+
+        const getParent = (categ) => {
+            if (categ.parent_id) {
+                current.push(categ.parent_id.id);
+                getParent(categ.parent_id);
+            }
+        };
+
+        for (const category of categories) {
+            current.push(category.id);
+            getParent(category);
+        }
+
+        return current;
+    }
+
+    get childPosCategIds() {
+        const current = [];
+        const categories = this.pos_categ_ids;
+
+        const getChild = (categ) => {
+            if (categ.child_id) {
+                for (const child of categ.child_id) {
+                    current.push(child.id);
+                    getChild(child);
+                }
+            }
+        };
+
+        for (const category of categories) {
+            current.push(category.id);
+            getChild(category);
+        }
+
+        return current;
+    }
+
+    get parent_pos_categ_ids() {
+        const categories = [];
+        let category = this.pos_categ_id;
+
+        while (category) {
+            categories.push(category.id);
+            category = category.parent_id;
+        }
+
+        return categories;
+    }
+
+    isPricelistItemUsable(item, date) {
+        return (
+            (!item.categ_id || this.parentCategoryIds.includes(item.categ_id.id)) &&
+            (!item.date_start || deserializeDate(item.date_start) <= date) &&
+            (!item.date_end || deserializeDate(item.date_end) >= date)
+        );
+    }
+
+    get_pricelist_rule(pricelist) {
+        const date = DateTime.now();
+        const applicableRules = [];
+
+        for (const item of pricelist.item_ids) {
+            if (!this.isPricelistItemUsable(item, date)) {
+                continue;
+            }
+
+            if (item.product_id && this.id === item.product_id.id) {
+                applicableRules.push(item);
+            } else if (
+                item._raw[item.id].product_tmpl_id &&
+                this._raw[this.id]?.product_tmpl_id === item._raw[item.id].product_tmpl_id
+            ) {
+                applicableRules.push(item);
+            } else if (
+                !item.product_id &&
+                !item.product_tmpl_id &&
+                !item._raw[item.id].product_tmpl_id &&
+                !item._raw[item.id].product_id
```

Comment:
To Joseph: It's possible that 'product.product' corresponding to `product_id` field is not load, so it has to be accessed from the `raw (or source)` data.